deviceplugin/usb.go

```diff
@@ -0,0 +1,169 @@
+	Product usbID `json:"product"`
+}
+
+type usbID uint16
+
+func (id *usbID) UnmarshalJSON(data []byte) error {
+	if string(data) == "null" || string(data) == `""` {
+		return nil
+	}
+	*id = usbID(binary.LittleEndian.Uint16(data))
+	return nil
+}
+
+func (id *usbID) String() string {
+	return fmt.Sprintf("%04x", int(*id))
+}
+
+// usbDevice represents a physical, tangible USB device.
+type usbDevice struct {
+	// Vendor is the USB Vendor ID of the device.
+	Vendor usbID `json:"vendor"`
+	// Product is the USB Product ID of the device.
+	Product usbID `json:"product"`
+	// Bus is the physical USB bus this device is located at.
+	Bus uint16 `json:"bus"`
+	// BusDevice is the location of the device on the Bus.
+	BusDevice uint16 `json:"busdev"`
+}
+
+func (dev *usbDevice) BusPath() (path string) {
+	return fmt.Sprintf(usbDevBus, dev.Bus, dev.BusDevice)
+}
+
+func enumerateUSBDevices() (specs []usbDevice, err error) {
+	allDevs, err := os.ReadDir(usbDevicesDir)
+	if err != nil {
+		return
+	}
+	for _, dev := range allDevs {
+		if !dev.IsDir() {
+			// There shouldn't be any raw files in this directory, but just in case.
+			continue
+		}
+		// Try to find the vendor ID file inside this device - this is a good indication that we're dealing with a device, not a bus.
+		vnd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirVendorIDFile)
+		if err != nil {
+			// We can't read the vendor file for some reason, it probably doesn't exist.
+			continue
+		}
+		prd, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirProductIDFile)
+		if err != nil {
+			continue
+		}
+
+		// The following two calls shouldn't fail.
+		busRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusFile)
+		if err != nil {
+			return specs, err
+		}
+		bus := binary.LittleEndian.Uint16(busRaw)
+		busLocRaw, err := os.ReadFile(dev.Name() + "/" + usbDevicesDirBusDevFile)
+		if err != nil {
+			return specs, err
+		}
+		busLoc := binary.LittleEndian.Uint16(busLocRaw)
+
+		spec := usbDevice{
+			Vendor:    usbID(binary.LittleEndian.Uint16(vnd)),
+			Product:   usbID(binary.LittleEndian.Uint16(prd)),
+			Bus:       bus,
+			BusDevice: busLoc,
+		}
+		specs = append(specs, spec)
+	}
+	return
+}
+
+func searchUSBDevices(devices *[]usbDevice, vendor usbID, product usbID) (devs []usbDevice, err error) {
+	for _, dev := range *devices {
+		if dev.Vendor == vendor && dev.Product == product {
+			devs = append(devs, dev)
+		}
+	}
+	return
+}
+
+func (gp *GenericPlugin) discoverUSB() (devices []device, err error) {
+	for _, group := range gp.ds.Groups {
+		paths := make([]string, len(group.UsbSpecs))
+		usbDevs, err := enumerateUSBDevices()
+		if err != nil {
+			return devices, err
+		}
+		for _, dev := range group.UsbSpecs {
+			matches, err := searchUSBDevices(&usbDevs, dev.Vendor, dev.Product)
+			if err != nil {
+				return nil, err
+			}
+			if len(matches) > 0 {
+				for _, match := range matches {
+					paths = append(paths, match.BusPath())
```

Comment:
Rather than appending here, we should be doing something like `paths[i] = match.BusPath()`
Alternatively, we can allocate the slice with a capacity but no length

Reply:
Arguably setting index is unnecessary, see b10fc167 (and above comment)

Reply:
On second thoughts, going to rework this a little

Reply:
See 6cde66f